### PR TITLE
nixos/spamassassin: Avoid network dependency on boot

### PR DESCRIPTION
**Pull request by pkern**
###### Motivation for this change

sa-update currently runs as part of the pre-start script of spamd. The network is not guaranteed to be online at that point and even if we were to depend on that, it makes the bootup brittle, as there is a reliance on SpamAssassin's update server as a startup dependency on boot.

Refactor the setup to move the pre-start script into its own oneshot unit (`spamd-init.service`). This allows to perform the setup task only once. Continuous updates are already done by `sa-update.service` triggered by `sa-update.timer`. Only run `sa-update` in case `/var/lib/spamassassin` is empty.

While we are on it, let sa-update.service depend on the network being online.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
  - None, because it's a change to a module.
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pkern** [comment]
@alyssais Hey Alyssa, would you mind having a look at this? It looks like you touched spamassassin last. ;-)

File: nixos/modules/services/mail/spamassassin.nix [hunk 1]
```diff
@@ -126,6 +126,8 @@ in
     };
 
     systemd.services.sa-update = {
+      wants = [ "network-online.target "];
+      after = [ "network-online.target "];
```
Comment (lukegb) [inline nixos/modules/services/mail/spamassassin.nix:130]:
```suggestion
      wants = [ "network-online.target" ];
      after = [ "network-online.target" ];
```

**Review by lukegb — APPROVED** [review]

File: nixos/modules/services/mail/spamassassin.nix (hunk 1 above)
Comment (pkern) [reply nixos/modules/services/mail/spamassassin.nix]: Sorry, I have no idea how this happened. Fixed.

**Comment by lukegb** [comment]
@ofborg eval

**Comment by pkern** [comment]
Gentle ping. :)

**Comment by alyssais** [comment]
Do we need the spamd-init service at all any more, then? All it does is create a state directory and do an update, so couldn’t it be replaced by a systemd `StateDirectory` and the update service?

**Comment by pkern** [comment]
@alyssais That's a good point. I think I made it work, PTAL. For spamd itself I did not see how you'd properly run it as non-root, though, so the systemd bits do escalation to root for starting and reloading but need to be set to user/group=spamd to ensure proper permissions on the state directory. I tested that initial startup and updates worked fine and did reload spamd without hanging.

While I was on it I also added sa-compile support post-update.

**Comment by pkern** [comment]
Although to be fair it might make sense to split out the sa-compile bit. It currently runs too often, i.e. not just when sa-update signals an update. But I guess I don't have a way to hook into that without using a script again or another service (which makes things more messy).

**Comment by pkern** [comment]
I re-added a script to only call sa-compile if updates have been installed. So now we are down to just a superfluous spamd reload in that case (because ExecStartPost cannot distinguish the two cases and we require elevation to do the systemctl invocation).

**Comment by alyssais** [comment]
> @alyssais That's a good point. I think I made it work, PTAL. For spamd
> itself I did not see how you'd properly run it as non-root, though, so
> the systemd bits do escalation to root for starting and reloading but
> need to be set to user/group=spamd to ensure proper permissions on the
> state directory. I tested that initial startup and updates worked fine
> and did reload spamd without hanging.

I was able to get it to work, but only if I gave it CAP_NET_BIND_SERVICE
so it could bind to port 783.  But I don't think it would be better than
what we have here, because with that approach spamd would be able to
bind to a low port at any time, whereas currently it can't do that after
priveleges have been dropped.  Ideally, of course, spamd would support
socket activation, but that is not the world we live in at the moment.

**Comment by alyssais** [comment]
Philipp Kern <notifications@github.com> writes:

> I re-added a script to only call sa-compile if updates have been
> installed.

Should we exit $rc at the end of that script?

> So now we are down to just a superfluous spamd reload in
> that case (because ExecStartPost cannot distinguish the two cases and
> we require elevation to do the systemctl invocation).

Could we use a path unit to do this instead of an ExecStartPost?
(systemd.path(5))

**Comment by alyssais** [comment]
>> @alyssais That's a good point. I think I made it work, PTAL. For spamd
>> itself I did not see how you'd properly run it as non-root, though, so
>> the systemd bits do escalation to root for starting and reloading but
>> need to be set to user/group=spamd to ensure proper permissions on the
>> state directory. I tested that initial startup and updates worked fine
>> and did reload spamd without hanging.
>
> I was able to get it to work, but only if I gave it CAP_NET_BIND_SERVICE
> so it could bind to port 783.  But I don't think it would be better than
> what we have here, because with that approach spamd would be able to
> bind to a low port at any time, whereas currently it can't do that after
> priveleges have been dropped.  Ideally, of course, spamd would support
> socket activation, but that is not the world we live in at the moment.

<https://github.com/systemd/systemd/pull/17655> would solve this, so we
might be able to revisit this in the future even if SpamAssassin doesn't
implement socket activation.

**Comment by pkern** [comment]
> Should we exit $rc at the end of that script?

sa-compile is only called when rc=0 and then the script returns the exit code of that. That's probably unfortunate in case sa-compile fails and systemd treats this as acceptable per `SuccessExitStatus`. Maybe it'd make sense to run sa-compile in its own unit somehow. This all seems messy, but "exit $rc" would not help. :/

> Could we use a path unit to do this instead of an ExecStartPost? (systemd.path(5))

I naively don't see how given that we'd need some notion that the updating and compiling is done. The last action sa-compile (currently) performs is copying to `/var/lib/spamassassin/<perlver>/<saver>/bases_body_0.pl`, but that seems odd to depend on.

**Comment by alyssais** [comment]
>> Should we exit $rc at the end of that script?
>
> sa-compile is only called when rc=0 and then the script returns the
> exit code of that. That's probably unfortunate in case sa-compile
> fails and systemd treats this as acceptable per
> `SuccessExitStatus`. Maybe it'd make sense to run sa-compile in its
> own unit somehow. This all seems messy, but "exit $rc" would not
> help. :/

I don't understand, then.  As I see it, there are the desirable
outcomes:

    sa-update exits 0 -> success
    sa-update exits 1 -> sa-compile exits 0 -> success
    sa-update exits 1 -> sa-compile something else -> fail
    sa-update exits something else -> fail

But the code you have here does not get this right, in the case where
sa-update exits 1, and sa-compile fails with exit 1, because then the
script will exit 1, and systemd will interpret that to be a success.

I think it should look like this (with the SuccessExitStatus line
dropped):

    ${pkgs.spamassassin}/bin/sa-update --verbose --gpghomedir=%S/spamassassin/sa-update-keys/ || [[ $? -eq 1 ]]
    ${pkgs.spamassassin}/bin/sa-compile

I think that's easier to read, too, because every step either returns
non-zero and exits, or returns 0 and carries on.  There's no need to
follow conditionals or think about different non-zero exit codes, except
where required to indicate that a 1 is also acceptable.

**Comment by alyssais** [comment]
> I think it should look like this (with the SuccessExitStatus line
> dropped):
>
>     ${pkgs.spamassassin}/bin/sa-update --verbose --gpghomedir=%S/spamassassin/sa-update-keys/ || [[ $? -eq 1 ]]
>     ${pkgs.spamassassin}/bin/sa-compile
>
> I think that's easier to read, too, because every step either returns
> non-zero and exits, or returns 0 and carries on.  There's no need to
> follow conditionals or think about different non-zero exit codes, except
> where required to indicate that a 1 is also acceptable.

Sorry, just realised I made an error of my own here, which was not
differentiating between the sa-update exit codes, so my suggestion
wouldn't work.

So I think your script is fine, but we should drop the SuccessExitStatus
line.  What do you think?

**Comment by pkern** [comment]
Hm, I guess in terms of states we want:

* sa-update rc=1 -> exit 0 (no update found, nothing to do; no benefit of plumbing through the 1 to systemd verbatim)
* sa-update rc=0 -> exit rc of sa-compile (which bash implicitly does for the last command in the script)
* any other sa-update rc gets passed through literally as an error

And then we remove `SuccessExitStatus` and we should be good. Or so I hope. I'll try to cook something along these lines up tomorrow.

**Comment by pkern** [comment]
PTAL. Testing both with updates available and subsequently no updates available worked fine for me.

**Comment by pkern** [comment]
Resolved the merge conflict. Is there anything else left to do for me? :)

**Comment by aanderse** [comment]
@pkern can you please remove the last 2 commits and do a rebase instead?

**Comment by pkern** [comment]
@aanderse Done.

**Comment by aanderse** [comment]
@alyssais @lukegb are either of you reviewing this module? Is it ready for merge?

**Comment by lukegb** [comment]
@aanderse I'm still happy with it; not sure if @alyssais has any outstanding commentary

**Comment by aanderse** [comment]
Well... in that case... "ping" to any motivated parties wanting to merge :smile:

**Pull request merged**